android/java/org/chromium/chrome/browser/crypto_wallet/fragments/ApproveTxBottomSheetDialogFragment.java

```diff
@@ -80,10 +83,24 @@ public static ApproveTxBottomSheetDialogFragment newInstance(
         mChainDecimals = 18;
     }
 
-    // TODO: This needs to be changed to something that broadcasts to all classes implementing
-    // ApprovedTxObserver.
-    public void setApprovedTxObserver(ApprovedTxObserver approvedTxObserver) {
-        mApprovedTxObserver = approvedTxObserver;
+    /**
+     * Add an observer to be notified when a transaction is approved or rejected.
+     */
+    public static void addApprovedTxObserver(ApprovedTxObserver approvedTxObserver) {
+        if (sApprovedTxObserver == null) sApprovedTxObserver = new ObserverList<>();
+        sApprovedTxObserver.addObserver(new WeakReference<>(approvedTxObserver));
+    }
+
+    /**
+     * Remove an observer of transactions changes.
+     */
+    public static void removeApprovedTxObserver(ApprovedTxObserver approvedTxObserver) {
+        if (sApprovedTxObserver == null) return;
+        for (int i = 0; i < sApprovedTxObserver.size(); ++i) {
+            if (sApprovedTxObserver.mObservers.get(i) == approvedTxObserver) {
```

Comment:
To compare observer reference use another `get` to obtain the underlying observer object.
`if (sApprovedTxObserver.mObservers.get(i).get() == approvedTxObserver) {`

Reply:
> Indeed it will be useless to have it in onStop so that's why I was preferring the approach to check the pending requests in onResume or something similar.

In that case it is better not having any observers at all, because of the activity problem. The only reason the other observers work is because they are singletons and binded to Mojo, so they live beyond UI lifecycles. With current structure, without singleton or `static` field there is no way to implement an observer.

Reply:
The [conditional](https://github.com/brave/brave-core/pull/12712#discussion_r836160219) change is required. Rest if the details has been discussed previously :)